# Assignment 6, Part 2


**Jane Programmer**

In this notebook, I am ingesting a fixed file that deals with precipitation measurements.  This data is used to spot trends in precipitation patterns. 

***Here, add some details about what the data includes and where to read more about it.  Include links, as your stakeholders might want to do their own reading.  You may find it useful to use copy/paste to save some time!***

In [1]:
import pandas as pd

I already opened the file in a text editor and discovered that the first __ rows are header/descriptive rows that I want to ignore and not read in. Let's see how close that gets us!

In [2]:
precip_df = pd.read_fwf('data/mctyprcp.txt', 
                        skiprows= # some number here -- peek at the file in a text editor to see how many,
                        header = # how do you indicate no header? )
precip_df.head()

0                1    2      3      4      5    6    7      8      9   \
0  AL  BIRMINGHAM    0  BHM  286.0 -324.0  159.0  1.0  1.0 -999.0 -999.0   
1  AL  HUNTSVILLE    0  HSV  387.0 -281.0  128.0  1.0  3.0 -999.0 -999.0   
2  AL  MOBILE        0  MOB  163.0 -557.0   75.0  8.0  1.0 -999.0 -999.0   
3  AL  MONTGOMERY    0  MGM  386.0 -253.0  211.0  8.0  3.0 -999.0 -999.0   
4  AK  ANCHORAGE     0  ANC    4.0  -61.0    2.0  2.0  0.0 -999.0 -999.0   

          10     11  
0  1548  -31   98.0  
1  2474  756  144.0  
2  1112 -681   62.0  
3  1086 -610   64.0  
4   172  -35   83.0

Let's look at the end of our data, also!

In [3]:
precip_df # how do you look at the end?

0                1    2      3      4      5    6    7      8      9   \
245  WY  CASPER        0  CPR  403.0  313.0  324.0  1.0  1.0 -999.0 -999.0   
246  WY  CHEYENNE      0  CYS  120.0   15.0   26.0  3.0  0.0 -999.0 -999.0   
247  WY  LANDER        0  LND  128.0    4.0   92.0  6.0  1.0 -999.0 -999.0   
248  WY  SHERIDAN      0  SHR   46.0  -54.0   17.0  9.0  0.0 -999.0 -999.0   
249  ND          OF DATA  NaN    NaN    NaN    NaN  NaN  NaN    NaN    NaN   

           10     11  
245  501  289  236.0  
246  205   12  106.0  
247  254   24  110.0  
248  187  -46   80.0  
249       NaN    NaN

This did a pretty good job as a first pass.  But, the last row of data (249) is the "end of data" and shouldn't be included.  Let's get rid of it!

In [4]:
precip_df = precip_df.drop(# some number goes here!)
precip_df.tail()

0                1    2      3      4      5    6    7      8      9   \
244  WI  WAUSAU        0  AUW  152.0  -40.0   60.0  8.0  2.0 -999.0 -999.0   
245  WY  CASPER        0  CPR  403.0  313.0  324.0  1.0  1.0 -999.0 -999.0   
246  WY  CHEYENNE      0  CYS  120.0   15.0   26.0  3.0  0.0 -999.0 -999.0   
247  WY  LANDER        0  LND  128.0    4.0   92.0  6.0  1.0 -999.0 -999.0   
248  WY  SHERIDAN      0  SHR   46.0  -54.0   17.0  9.0  0.0 -999.0 -999.0   

           10     11  
244  700  309  179.0  
245  501  289  236.0  
246  205   12  106.0  
247  254   24  110.0  
248  187  -46   80.0

Also, the second column ("1") has two values in it.  We can split this column easily because we know that the last character (position -1) is the "pflag" and that everything else belongs to the city name.  Let's create two new columns, `city` and `pflag`, which are the result of splitting the combined column.

In [5]:
precip_df['city'] = precip_df[1].str[:-1]  # We're giving you this one for free...
precip_df['pflag'] = precip_df[1].str[-1]  # But you'll want to remember this later!

Did this work?

In [6]:
precip_df.head()

0                1    2      3      4      5    6    7      8      9  \
0  AL  BIRMINGHAM    0  BHM  286.0 -324.0  159.0  1.0  1.0 -999.0 -999.0   
1  AL  HUNTSVILLE    0  HSV  387.0 -281.0  128.0  1.0  3.0 -999.0 -999.0   
2  AL  MOBILE        0  MOB  163.0 -557.0   75.0  8.0  1.0 -999.0 -999.0   
3  AL  MONTGOMERY    0  MGM  386.0 -253.0  211.0  8.0  3.0 -999.0 -999.0   
4  AK  ANCHORAGE     0  ANC    4.0  -61.0    2.0  2.0  0.0 -999.0 -999.0   

          10     11            city pflag  
0  1548  -31   98.0  BIRMINGHAM         0  
1  2474  756  144.0  HUNTSVILLE         0  
2  1112 -681   62.0  MOBILE             0  
3  1086 -610   64.0  MONTGOMERY         0  
4   172  -35   83.0  ANCHORAGE          0

Great, it did!  Let's remove column 1, since we've captured its data elsewhere.

In [7]:
precip_df = precip_df.drop(columns=[ # some number goes here -- which column are we dropping?])
precip_df.head()

0    2      3      4      5    6    7      8      9         10     11  \
0  AL  BHM  286.0 -324.0  159.0  1.0  1.0 -999.0 -999.0  1548  -31   98.0   
1  AL  HSV  387.0 -281.0  128.0  1.0  3.0 -999.0 -999.0  2474  756  144.0   
2  AL  MOB  163.0 -557.0   75.0  8.0  1.0 -999.0 -999.0  1112 -681   62.0   
3  AL  MGM  386.0 -253.0  211.0  8.0  3.0 -999.0 -999.0  1086 -610   64.0   
4  AK  ANC    4.0  -61.0    2.0  2.0  0.0 -999.0 -999.0   172  -35   83.0   

             city pflag  
0  BIRMINGHAM         0  
1  HUNTSVILLE         0  
2  MOBILE             0  
3  MONTGOMERY         0  
4  ANCHORAGE          0

Uh oh, it looks like column 10 also needs to be split!  We can do this again by position, first 4 (up to but not including position 5) and last 4 (starting with position 5 and continuing until the end):

In [8]:
precip_df['year_to_date_amt'] = precip_df[10].str # add some indexing here... how do you say "up to 5" ?
precip_df['year_to_date_dep_nml'] = precip_df[10].str # add some indexing here ... how do you say "5 and up" ?

precip_df = precip_df.drop # what are we dropping?  Add some code here!
precip_df.head()

0    2      3      4      5    6    7      8      9     11  \
0  AL  BHM  286.0 -324.0  159.0  1.0  1.0 -999.0 -999.0   98.0   
1  AL  HSV  387.0 -281.0  128.0  1.0  3.0 -999.0 -999.0  144.0   
2  AL  MOB  163.0 -557.0   75.0  8.0  1.0 -999.0 -999.0   62.0   
3  AL  MGM  386.0 -253.0  211.0  8.0  3.0 -999.0 -999.0   64.0   
4  AK  ANC    4.0  -61.0    2.0  2.0  0.0 -999.0 -999.0   83.0   

             city pflag year_to_date_amt year_to_date_dep_nml  
0  BIRMINGHAM         0            1548                   -31  
1  HUNTSVILLE         0            2474                   756  
2  MOBILE             0            1112                  -681  
3  MONTGOMERY         0            1086                  -610  
4  ANCHORAGE          0            172                    -35

Now, all we have to do are supply column names and maybe re-order things so that the city and pflag come as the second and third columns, respectively:

In [9]:
precip_df.rename(columns={0:'state',
                          2: "call_ltrs",
                          3: "month_tot_amt",
                          4: "month_dep_nml",
                          5: "24_hr_max",
                          6: "days_gt_1",
                          7: "days_gt_50",
                          8: "season_to_date",
                          9: "season_dep_nml",
                          11: "year_to_date_pct_nml"},inplace=True)

precip_df = precip_df[["state",
                       "city",
                       "pflag",
# Add the columns here, in the order they appear in the original data...
                       "year_to_date_dep_nml",
                       "year_to_date_pct_nml"]]

In [10]:
precip_df.head()

state            city pflag call_ltrs  month_tot_amt  month_dep_nml  \
0    AL  BIRMINGHAM         0       BHM          286.0         -324.0   
1    AL  HUNTSVILLE         0       HSV          387.0         -281.0   
2    AL  MOBILE             0       MOB          163.0         -557.0   
3    AL  MONTGOMERY         0       MGM          386.0         -253.0   
4    AK  ANCHORAGE          0       ANC            4.0          -61.0   

   24_hr_max  days_gt_1  days_gt_50  season_to_date  season_dep_nml  \
0      159.0        1.0         1.0          -999.0          -999.0   
1      128.0        1.0         3.0          -999.0          -999.0   
2       75.0        8.0         1.0          -999.0          -999.0   
3      211.0        8.0         3.0          -999.0          -999.0   
4        2.0        2.0         0.0          -999.0          -999.0   

  year_to_date_amt year_to_date_dep_nml  year_to_date_pct_nml  
0            1548                   -31                  98.0  
1            2474                   756                 144.0  
2            1112                  -681                  62.0  
3            1086                  -610                  64.0  
4            172                    -35                  83.0

Now let's identify the city with the highest and lowest `year_to_date_pct_nml`.  We'll sort first, then we can just look at the first and the last row.

In [11]:
precip_df = precip_df # how do we sort by a column value?

What city has the lowest YTD percent normal precipitation?

In [12]:
precip_df # how can you use iloc here to get the first (0th) row?

state                               SC
city                    CHARLESTON    
pflag                                0
call_ltrs                          CHS
month_tot_amt                       70
month_dep_nml                     -330
24_hr_max                           29
days_gt_1                            6
days_gt_50                           0
season_to_date                    -999
season_dep_nml                    -999
year_to_date_amt                 306 -
year_to_date_dep_nml               827
year_to_date_pct_nml                27
Name: 189, dtype: object

Charleston, SC only has 27% of its year to date normal.  That's dry!

And the city with the highest percent of its normal precipitation?

In [13]:
precip_df # How can you use iloc here to get the last row?

state                               AK
city                    BARROW        
pflag                                0
call_ltrs                          BRW
month_tot_amt                       69
month_dep_nml                       60
24_hr_max                           18
days_gt_1                            3
days_gt_50                           0
season_to_date                    -999
season_dep_nml                    -999
year_to_date_amt                 232  
year_to_date_dep_nml               199
year_to_date_pct_nml               703
Name: 5, dtype: object

Barrow, Alaska is very very wet (perhaps with snow?) at over 7 times its normal precipitation!  Wow!

We've done a lot of work to re-format our data.  Let's write this neatned up dataset to file!

In [14]:
precip_df.to_csv('data/precipitation_data.csv') # Or wherever you save your data!